[GOBBLIN-200] Add cleanable state store dataset

jack-moseley opened the pull request:
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-200


### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

Create wrappers for `DatasetStoreDataset` and `DatasetStoreDatasetFinder` to make them Cleanable

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Created `CleanableDatasetStoreDatasetTest`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

jack-moseley commented:
@ibuenros please review

yukuai518 commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/TimestampedDatasetStateStoreVersion.java:
```diff
@@ -0,0 +1,53 @@
+    return this.version.equals(otherAsDateTime.version) ? 0 : this.version.compareTo(otherAsDateTime.version);
+  }
+
+  @Override
```
You can remove this if the behavior is the same as parent class

jack-moseley replied:
Removing this triggers findbugs warning, which recommends explicitly overriding even if behaviour is the same.

yukuai518 commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/dataset/CleanableDatasetStoreDataset.java:
```diff
@@ -0,0 +1,62 @@
+ * A cleanable {@link DatasetStoreDataset}
+ */
+@Data
+public class CleanableDatasetStoreDataset extends ModificationTimeDataset {
```
We probably don't want to assume the CleanableDatasetStoreDataset is always using Timebased version finder.

yukuai518 commented on gobblin-metastore/src/main/java/org/apache/gobblin/metastore/DatasetStoreDatasetFinder.java:
```diff
@@ -76,7 +77,7 @@ private StateStorePredicate buildPredicate() {
 
   @Override
-  public List<DatasetStoreDataset> findDatasets() throws IOException {
+  public List<Dataset> findDatasets() throws IOException {
```
Similar problem.

jack-moseley commented:
@yukuai518 refactored based on your comments:

- Removed CleanableDataset dependency on FileSystem
- Changed CleanableDatasetStoreDataset to inherit directly from CleanableDataset
- Changed CleanableDatasetStoreDataset to be an abstract base class where subclasses can choose which VersionFinder/Policy to use
- Added DatasetStateStoreVersion

ibuenros commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/dataset/CleanableDatasetStoreDataset.java:
```diff
@@ -0,0 +1,59 @@
+
+    Collections.sort(versions, Collections.reverseOrder());
+
+    Collection<T> deletableVersions = this.getVersionSelectionPolicy().listSelectedVersions(versions);
```
Shouldn't this be the other way around? `VersionSelectionPolicy` selects the versions to keep.

jack-moseley replied:
It could be used as versions to delete as well, and `MultiVersionCleanableDatasetBase` does it this way.

ibuenros commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/DatasetStateStoreVersion.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * {@link DatasetVersion} that has a {@link DatasetStateStoreEntryManager}
+ */
+public interface DatasetStateStoreVersion extends DatasetVersion {
```
What is the point of this class?

jack-moseley replied:
Allows `CleanableDatasetStoreDataset` to call getEntry()

ibuenros commented:
+1
